src/software/new_geom/polygon_test.cpp

```diff
@@ -0,0 +1,218 @@
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(0, 0.5)));  // on a "right" edge, see NOTE on Polygon::contains
+    EXPECT_TRUE(intersecting_poly.contains(Point(0, 1)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(1, 1.5)));
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(0, 2)));  // on a top edge, see NOTE on Polygon::contains
+    EXPECT_TRUE(intersecting_poly.contains(Point(1, 1)));
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(1, 2)));  // on a top edge, see NOTE on Polygon::contains
+    EXPECT_TRUE(intersecting_poly.contains(Point(0, 1.5)));
+}
+
+TEST(PolygonTest, test_complex_self_intersecting_polygon_contains)
+{
+    /*
+     *  2            *-------*
+     *               |       |
+     *  1            *-------*
+     *               |
+     *  0            *
+     *               |
+     *  -1   *-------*
+     *       |       |
+     *  -2   *-------*
+     *
+     *      -2       0       2
+     */
+    // Self intersecting polygon, each asterisk on the diagram is a point making up the
+    // polygon
+    Polygon intersecting_poly{{0.0f, 0.0f},   {0.0f, 2.0f},   {2.0f, 2.0f},
+                              {2.0f, 1.0f},   {0.0f, 1.0f},   {-0.0f, -2.0f},
+                              {-2.0f, -2.0f}, {-2.0f, -1.0f}, {0.0f, -1.0f}};
+
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point()));  // on a "right" edge, see NOTE on Polygon::contains
+    EXPECT_FALSE(intersecting_poly.contains(Point(2, 0)));
+    EXPECT_FALSE(intersecting_poly.contains(Point(0.5, 0.5)));
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(0, 0.5)));  // on a "right" edge, see NOTE on Polygon::contains
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(0, 2)));  // on a top edge, see NOTE on Polygon::contains
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(1, 2)));  // on a top edge, see NOTE on Polygon::contains
+    EXPECT_FALSE(intersecting_poly.contains(Point(-2, 0)));
+    EXPECT_FALSE(intersecting_poly.contains(Point(-0.5, -0.5)));
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(0, -0.5)));  // on a "right" edge, see NOTE on Polygon::contains
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(0, -2)));  // on a top edge, see NOTE on Polygon::contains
+    EXPECT_FALSE(intersecting_poly.contains(Point(0, -1)));
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(-1, -1)));  // on a "top" edge, see NOTE on Polygon::contains
+    EXPECT_FALSE(intersecting_poly.contains(
+        Point(0, -1.5)));  // on a "right" edge, see NOTE on Polygon::contains
+    EXPECT_TRUE(intersecting_poly.contains(Point(-1, -2)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(0, 1)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(1, 1.5)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(1, 1)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(0, 1.5)));
+    EXPECT_TRUE(intersecting_poly.contains(Point(-1, -1.5)));
+}
```

Comment:
One last test to suggest:
```
XXXXXXXXXXXXXXXXXXXXXXXXXXXX
X                          XX
X                           X
X         XXXXXXXX          X
X         X     XX          X
X          XX   X           X
X            XXXXX          X
X          XXX   XXXXX     XX
XXXXXXXXXXXX         XXXXXXX
```
(apologies for the horrid ascii drawing, just wanted to quickly convey the idea).

The impetus for suggesting this is I just want to be sure we're being explicit what's "contained" in a polygon like this.

Reply:
Added a test for this. Such loops are treated as "holes" and any points in them aren't contained.

Reply:
Sorry if I'm just missing it, but I don't see this test here?

Reply:
Check line 174 of polygon_test.cpp, something weird was going on - it wasnt committed in my previous commit for some reason. Should be there now.